src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java

```diff
@@ -701,7 +717,7 @@ protected void doXContentBody(XContentBuilder builder, boolean includeDefaults,
         if (includeDefaults || fieldType.stored() != defaultFieldType.stored()) {
             builder.field("store", fieldType.stored());
         }
-        if (includeDefaults || hasDocValues() != Defaults.DOC_VALUES) {
+        if (includeDefaults || hasDocValues() != defaultDocValues()) {
```

Comment:
fieldType is not mutable here.  It is frozen in the ctor of AbstractFieldMapper.  This needs to be overridable by subclasses so e.g. geo mapper can use false instead of basing on the field type.